### DISCUSSION TITLE
Tips for speeding up SMARTS matching on large molecules?

### DISCUSSION THREAD

**Discussion by j-wags** [rdkit/rdkit#3939] [Q&A] [answered]
I'm working on applying protein force fields in the SMIRNOFF format. This requires doing SMARTS searches on a large RDMol, which takes a while. I was wondering if any folks here had tips for improving the runtime of the SMARTS matching code in the "thousands-of-SMARTS-and-protein-sized-molecule" regime.

[Please find a reproducing case here](https://github.com/rdkit/rdkit/files/6152959/biopolymer-smarts-example.tar.gz) (requires `conda install -c conda-forge rdkit tqdm`)

Current timing on my computer is RDKit: 1 minute, 23 seconds and OpenEye: 1 minute, 34 seconds

So far, I thought of multithreading the search (is RDKit's SMARTS matching thread-safe?) and preprocessing the SMARTS list to remove identical substructures. Does anyone have other possible speed-ups that could be implemented?

h/t @SimonBoothroyd who made this minimal example

**Comment by greglandrum** [comment, marked as answer]
This isn't connected to the SMARTS matching, but one way to reduce the runtime of that code dramatically is to stop copying the molecule and re-sanitizing it in the find_smarts_matches() function. That's an expensive operation and you are repeating it for every SMARTS.
If I comment out these three lines:
```
    # rd_molecule = Chem.Mol(rd_molecule)

    # Chem.SanitizeMol(rd_molecule, Chem.SANITIZE_ALL ^ Chem.SANITIZE_SETAROMATICITY)
    # Chem.SetAromaticity(rd_molecule, Chem.AromaticityModel.AROMATICITY_MDL)
```
it reduces the runtime from 36s to 6s on my machine.

**Reply by greglandrum** [reply]
As for thread safety, this piece of the documentation isn't 100% complete, but it is accurate:
http://www.rdkit.org/docs/RDKit_Book.html#thread-safety-and-the-rdkit

And, yes, substructure matching is thread safe.

**Comment by j-wags** [comment]
Thank you, @greglandrum! I must have had tunnel vision to miss the cost of re-sanitizing. That's a huge time savings, and we can implement an equivalent in the OpenFF Toolkit code. 

Apologies for not seeing the thread safety section in the docs. I should have looked more closely for that. Our substructures shouldn't have recursive elements so I'll go ahead with implementing multithreading.

I'll be sure to mention your help when we roll out official support for biopolymers at OpenFF!
